frontend/src/Frontend/Storage/Class.hs

```diff
@@ -0,0 +1,299 @@
+{-# LANGUAGE CPP #-}
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE DefaultSignatures #-}
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+module Frontend.Storage.Class
+  ( localStorage
+  , sessionStorage
+  , getItemStorage
+  , setItemStorage
+  , removeItemStorage
+  , removeKeyUniverse
+  , dumpLocalStorage
+  , backupLocalStorage
+  , restoreLocalStorage
+  , restoreLocalStorageDump
+  , getCurrentVersion
+  , StoreType (..)
+  , HasStorage(..)
+  -- Versioning Stuff
+  , StorageVersioner(..)
+  , StoreKeyMetaPrefix(..)
+  , VersioningError(..)
+  , StorageVersion
+  ) where
+
+import Control.Monad.Reader
+import Data.Aeson (FromJSON, ToJSON)
+import Data.Constraint.Extras (Has, Has', has)
+import Data.Constraint.Forall (ForallF)
+import Data.Dependent.Map (DMap, DSum((:=>)))
+import Data.Dependent.Sum.Orphans ()
+import Data.Functor.Identity (Identity(Identity))
+import Data.GADT.Compare (GCompare)
+import Data.GADT.Show (GShow,gshow)
+import Data.Proxy (Proxy)
+import Data.Some (Some(Some))
+import Data.Text (Text)
+import Data.Universe.Some (UniverseSome, universeSome)
+import Frontend.Foundation
+import Numeric.Natural (Natural)
+import Obelisk.Route.Frontend
+
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString.Lazy as BL
+import qualified Data.Dependent.Map as DMap
+import qualified Data.Text as T
+import qualified Data.Text.Encoding as T
+
+{-- Notes for migration
+ * We can't close off the keys to a single GADT because that would mean that
+   desktop and web need to share the same storage keys, which will not work.
+   Currently Desktop has the root BIP key on top of the rest of the storage.
+ * The backup/restore process for the user is going to be interesting. Because we don't
+   want to export the whole store because I don't think that we want to export the
+   encrypted keys. It's not quite what we want for the versioning process so we are not
+   going to get it "for free" sadly.
+ * The current storage type locks things up in JSM which makes it really tricky to test.
+   We could run the frontend storage tests in ghcjs but then we'd be stuck not being able to
+   test the desktop migrations. We could parameterise Storage with a type parameter for the
+   inner monad, but then HasStorage needs to become a MTPC and stuff gets weird for the rest
+   of the app. Really, we just want to be able to test the derived functions from the algebra
+   and we don't want "testability" stuff bleeding out into the app.
+   For this reason, I'm going to change Storage to a Free Monad and HasStorage / StorageT just
```

Comment:
Not so free anymore

Reply:
True. We could probably just delete this comment all together now.